File: packages/contracts/src/agent-events.ts
import { z } from 'zod';

const eventBase = z.object({
  requestId: z.string().uuid(),
  traceId: z.string().min(1),
  seq: z.int().nonnegative(),
  occurredAt: z.string().datetime(),
}).strict();

const citationLocation = z.object({
  page: z.int().positive().optional(),
  slide: z.int().positive().optional(),
  sheet: z.string().min(1).optional(),
  cellRange: z.string().min(1).optional(),
}).strict();

const retrievalPathSummary = z.object({
  path: z.enum(['vector', 'lexical']),
  spaceId: z.string().min(1),
  candidatesReturned: z.int().nonnegative(),
  candidatesAfterAcl: z.int().nonnegative().optional(),
  error: z.string().optional(),
}).strict();

const retrievalSummary = z.object({
  query: z.string(),
  originalQuery: z.string(),
  contextualized: z.boolean(),
  attempt: z.int().positive(),
  vectorTopK: z.int().nonnegative(),
  lexicalTopK: z.int().nonnegative(),
  rrfK: z.int().nonnegative(),
  rrfTopK: z.int().nonnegative(),
  rerankTopK: z.int().nonnegative(),
  rerankerEnabled: z.boolean(),
  rerankerFailed: z.boolean().optional(),
  paths: z.array(retrievalPathSummary),
  rrfCandidateCount: z.int().nonnegative(),
  finalCandidateCount: z.int().nonnegative(),
  embeddingModel: z.string(),
  embeddingVersion: z.string(),
  rerankerModel: z.string(),
  rerankerVersion: z.string(),
  elapsedMs: z.number().nonnegative(),
}).strict();

export const agentEventSchema = z.discriminatedUnion('type', [
  eventBase.extend({ type: z.literal('run.started') }).strict(),
  eventBase.extend({
    type: z.literal('run.status'),
    status: z.enum(['understanding', 'retrieving', 'ranking', 'answering']),
  }).strict(),
  eventBase.extend({
    type: z.literal('text.delta'),
    text: z.string().min(1),
  }).strict(),
  eventBase.extend({
    type: z.literal('citation'),
    citationId: z.string().uuid(),
    chunkId: z.string().uuid(),
    documentId: z.string().uuid(),
    title: z.string().min(1),
    snippet: z.string(),
    location: citationLocation,
  }).strict(),
  eventBase.extend({
    type: z.literal('retrieval.summary'),
    summary: retrievalSummary,
  }).strict(),
  eventBase.extend({
    type: z.literal('run.completed'),
    finishReason: z.enum(['stop', 'cancelled']),
  }).strict(),
  eventBase.extend({
    type: z.literal('run.failed'),
    code: z.string().min(1),
    message: z.string().min(1),
    retryable: z.boolean(),
  }).strict(),
]);

export type AgentEvent = z.infer<typeof agentEventSchema>;

export const chatHistoryMessageSchema = z.object({
  role: z.enum(['user', 'assistant']),
  content: z.string().min(1),
}).strict();

const historySummarySchema = z.string()
  .refine((value) => Array.from(value).length <= 4_000, {
    message: 'history summary must be at most 4000 Unicode code points',
  })
  .refine((value) => {
    const lines = value.split(/\r?\n/);
    const topics = lines.slice(1);
    const isCanonicalSummary =
      lines[0] === '此前已授权的用户话题：' &&
      topics.length > 0 &&
      topics.every((topic) => topic.startsWith('- '));
    return !isCanonicalSummary || (
      topics.length <= 20 && topics.every((topic) => Array.from(topic.slice(2)).length <= 240)
    );
  }, {
    message: 'canonical history summaries allow at most 20 topics of 240 Unicode code points each',
  })
  .default('');

export const runRequestSchema = z.object({
  requestId: z.string().uuid(),
  traceId: z.string().min(1),
  actorId: z.string().min(1),
  question: z.string().trim().min(1).max(8000),
  selectedSpaceIds: z.array(z.string().uuid()).max(100),
  aclSnapshot: z.record(z.string(), z.unknown()).default({}),
  sessionId: z.string().max(120).optional(),
  history: z.array(chatHistoryMessageSchema).default([]),
  historySummary: historySummarySchema,
}).strict();

export type RunRequest = z.infer<typeof runRequestSchema>;
export type ChatHistoryMessage = z.infer<typeof chatHistoryMessageSchema>;

export const chatRequestSchema = runRequestSchema.extend({
  history: z.array(chatHistoryMessageSchema).default([]),
  historySummary: historySummarySchema,
}).strict();

export type ChatRequest = z.infer<typeof chatRequestSchema>;

export type CitationLocation = z.infer<typeof citationLocation>;
export type RetrievalSummary = z.infer<typeof retrievalSummary>;
export type RetrievalPathSummary = z.infer<typeof retrievalPathSummary>;

File: services/ai/src/rag_ai/contracts/agent_events.py
from datetime import datetime
from typing import Annotated, Any, Literal, TypeAlias
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class EventBase(StrictModel):
    requestId: UUID
    traceId: str = Field(min_length=1)
    seq: int = Field(ge=0)
    occurredAt: datetime


class RunStarted(EventBase):
    type: Literal["run.started"]


class RunStatus(EventBase):
    type: Literal["run.status"]
    status: Literal["understanding", "retrieving", "ranking", "answering"]


class TextDelta(EventBase):
    type: Literal["text.delta"]
    text: str = Field(min_length=1)


class CitationLocation(StrictModel):
    page: int | None = Field(default=None, ge=1)
    slide: int | None = Field(default=None, ge=1)
    sheet: str | None = Field(default=None, min_length=1)
    cellRange: str | None = Field(default=None, min_length=1)

    @model_validator(mode="before")
    @classmethod
    def reject_explicit_nulls(cls, value: object) -> object:
        location_fields = ("page", "slide", "sheet", "cellRange")
        if isinstance(value, dict) and any(
            field in value and value[field] is None for field in location_fields
        ):
            raise ValueError("citation location fields cannot be null")
        return value


class Citation(EventBase):
    type: Literal["citation"]
    citationId: UUID
    chunkId: UUID
    documentId: UUID
    title: str = Field(min_length=1)
    snippet: str
    location: CitationLocation


class RetrievalPathSummary(StrictModel):
    path: Literal["vector", "lexical"]
    spaceId: str = Field(min_length=1)
    candidatesReturned: int = Field(ge=0)
    candidatesAfterAcl: int | None = Field(default=None, ge=0)
    error: str | None = Field(default=None)


class RetrievalSummary(EventBase):
    type: Literal["retrieval.summary"]
    summary: dict[str, Any]


class RunCompleted(EventBase):
    type: Literal["run.completed"]
    finishReason: Literal["stop", "cancelled"]


class RunFailed(EventBase):
    type: Literal["run.failed"]
    code: str = Field(min_length=1)
    message: str = Field(min_length=1)
    retryable: bool


AgentEvent: TypeAlias = Annotated[
    RunStarted
    | RunStatus
    | TextDelta
    | Citation
    | RetrievalSummary
    | RunCompleted
    | RunFailed,
    Field(discriminator="type"),
]


class RunRequest(StrictModel):
    requestId: UUID
    traceId: str = Field(min_length=1)
    actorId: str = Field(min_length=1)
    question: str = Field(min_length=1, max_length=8000)
    selectedSpaceIds: list[UUID] = Field(max_length=100)
    aclSnapshot: dict[str, Any] = Field(default_factory=dict)
    sessionId: str | None = Field(default=None, max_length=120)

    @field_validator("question", mode="before")
    @classmethod
    def strip_question(cls, value: object) -> object:
        return value.strip() if isinstance(value, str) else value

    @field_validator("aclSnapshot", mode="before")
    @classmethod
    def default_acl_snapshot(cls, value: object) -> object:
        return value if isinstance(value, dict) else {}


class ChatHistoryMessage(StrictModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1)


class ChatRequest(RunRequest):
    history: list[ChatHistoryMessage] = Field(default_factory=list)
    historySummary: str = Field(default="", max_length=4000)
